File: crates/kaleido-core/bindings/swift-probe/Probe.swift
import Foundation

func probeProtocolVersion() -> String {
    protocolVersion()
}

func probeCanonicalGraph(
    command: CommandEnvelope,
    projection: ProjectionEnvelope,
    error: CanonicalError?,
    effects: [StateEffect]
) -> ProjectionEnvelope {
    bindingProbe(
        command: command,
        projection: projection,
        error: error,
        effects: effects
    )
}

private final class ProjectionProbeSink: ProjectionProbeCallback {
    func onProjection(projection: ProjectionEnvelope) {
        _ = consumeProjectionKey(projection.key)
    }

    func onError(error: CanonicalError) {
        _ = error
    }
}

func probeSubscription(
    projection: ProjectionEnvelope,
    error: CanonicalError
) {
    let subscription = ProjectionSubscriptionProbe()
    subscription.subscribe(
        callback: ProjectionProbeSink(),
        projection: projection,
        error: error
    )
    subscription.unsubscribe()
}

func probeFallibleCall(error: CanonicalError) -> CanonicalError? {
    do {
        try fallibleBindingProbe(shouldFail: true, error: error)
        return nil
    } catch let BindingProbeError.Canonical(error: canonicalError) {
        return canonicalError
    } catch {
        return nil
    }
}

func probeAsyncCall(ack: CommandAck) async -> CommandAck {
    await asyncBindingProbe(ack: ack)
}

func probeProjectionContracts(
    hostId: HostId,
    projectId: ProjectId,
    sessionId: SessionId,
    workflowId: WorkflowId,
    runtimeId: ProviderRuntimeId,
    cursor: Cursor,
    error: CanonicalError
) -> [ProjectionSubscribeAck] {
    let keys: [ProjectionKey] = [
        .projectIndex(hostId: hostId),
        .sessionIndex(projectId: projectId),
        .transcript(sessionId: sessionId),
        .liveActivity(sessionId: sessionId),
        .inputQueue(sessionId: sessionId),
        .attentionInbox(hostId: hostId),
        .workflowBoard(workflowId: workflowId),
        .runtimeCapability(hostId: hostId, runtimeId: runtimeId),
    ]
    let outcomes: [ProjectionSubscribeOutcome] = [
        .resumed(fromCursor: cursor),
        .currentFollows(currentCursor: cursor),
        .rejected(error: error),
    ]

    return keys.enumerated().map { index, key in
        let request = ProjectionSubscribe(key: key, since: cursor)
        _ = consumeProjectionKey(request.key)
        let ack = ProjectionSubscribeAck(
            key: request.key,
            outcome: outcomes[index % outcomes.count]
        )
        _ = consumeProjectionOutcome(ack.outcome)
        return ack
    }
}

func probeMobileIngress(
    deviceId: DeviceId,
    body: Command,
    contentRef: ContentRef,
    error: CanonicalError
) -> String {
    let actor: Actor = .human(deviceId: deviceId)
    let command = DeviceCommandRequest(idempotencyKey: "probe", ttlMs: 1, body: body)
    let write = ContentWriteRequest(
        contentKind: .plainText,
        byteLen: contentRef.byteLen,
        digest: contentRef.digest
    )
    let responses: [ContentWriteResponse] = [
        .stored(contentRef: contentRef),
        .rejected(error: error),
    ]

    return consumeActor(actor)
        + command.idempotencyKey
        + String(describing: write.contentKind)
        + responses.map(consumeContentWriteResponse).joined()
}

func probeRuntimeCapabilityScope(view: RuntimeCapabilityView) -> String {
    "\(view.hostId.value):\(view.runtimeId.value)"
}

private func consumeActor(_ actor: Actor) -> String {
    switch actor {
    case let .human(deviceId):
        return deviceId.value
    case let .workflow(workflowId):
        return workflowId.value
    case .broker:
        return "broker"
    }
}

private func consumeProjectionKey(_ key: ProjectionKey) -> String {
    switch key {
    case let .projectIndex(hostId):
        return hostId.value
    case let .sessionIndex(projectId):
        return projectId.value
    case let .transcript(sessionId):
        return sessionId.value
    case let .liveActivity(sessionId):
        return sessionId.value
    case let .inputQueue(sessionId):
        return sessionId.value
    case let .attentionInbox(hostId):
        return hostId.value
    case let .workflowBoard(workflowId):
        return workflowId.value
    case let .runtimeCapability(hostId, runtimeId):
        return "\(hostId.value):\(runtimeId.value)"
    }
}

private func consumeProjectionOutcome(_ outcome: ProjectionSubscribeOutcome) -> String {
    switch outcome {
    case let .resumed(fromCursor):
        return String(fromCursor.seq)
    case let .currentFollows(currentCursor):
        return String(currentCursor.seq)
    case let .rejected(error):
        return String(describing: error.code)
    }
}

private func consumeContentWriteResponse(_ response: ContentWriteResponse) -> String {
    switch response {
    case let .stored(contentRef):
        return contentRef.digest
    case let .rejected(error):
        return String(describing: error.code)
    }
}

File: crates/kaleido-core/bindings/kotlin-probe/src/main/kotlin/Probe.kt
package probe

import uniffi.kaleido_core.BindingProbeException
import uniffi.kaleido_core.ProjectionProbeCallback
import uniffi.kaleido_core.ProjectionSubscriptionProbe
import uniffi.kaleido_core.asyncBindingProbe
import uniffi.kaleido_core.bindingProbe
import uniffi.kaleido_core.fallibleBindingProbe
import uniffi.kaleido_core.MobileQuestionAnswer
import uniffi.kaleido_core.protocolVersion
import uniffi.kaleido_proto.Actor
import uniffi.kaleido_proto.CanonicalError
import uniffi.kaleido_proto.CommandAck
import uniffi.kaleido_proto.Command
import uniffi.kaleido_proto.CommandEnvelope
import uniffi.kaleido_proto.ContentKind
import uniffi.kaleido_proto.ContentRef
import uniffi.kaleido_proto.ContentWriteRequest
import uniffi.kaleido_proto.ContentWriteResponse
import uniffi.kaleido_proto.Cursor
import uniffi.kaleido_proto.DeviceCommandRequest
import uniffi.kaleido_proto.DeviceId
import uniffi.kaleido_proto.HostId
import uniffi.kaleido_proto.ProjectId
import uniffi.kaleido_proto.ProviderRuntimeId
import uniffi.kaleido_proto.ProjectionEnvelope
import uniffi.kaleido_proto.ProjectionKey
import uniffi.kaleido_proto.ProjectionSubscribe
import uniffi.kaleido_proto.ProjectionSubscribeAck
import uniffi.kaleido_proto.ProjectionSubscribeOutcome
import uniffi.kaleido_proto.RuntimeCapabilityView
import uniffi.kaleido_proto.SessionId
import uniffi.kaleido_proto.StateEffect
import uniffi.kaleido_proto.WorkflowId
import uniffi.kaleido_proto.QuestionAnswer
import uniffi.kaleido_proto.QuestionPrompt

fun probeProtocolVersion(): String = protocolVersion()

fun probeCanonicalGraph(
    command: CommandEnvelope,
    projection: ProjectionEnvelope,
    error: CanonicalError?,
    effects: List<StateEffect>,
): ProjectionEnvelope = bindingProbe(command, projection, error, effects)

private class ProjectionProbeSink : ProjectionProbeCallback {
    override fun onProjection(projection: ProjectionEnvelope) {
        consumeProjectionKey(projection.key)
    }

    override fun onError(error: CanonicalError) {
        checkNotNull(error)
    }
}

fun probeSubscription(
    projection: ProjectionEnvelope,
    error: CanonicalError,
) {
    val subscription = ProjectionSubscriptionProbe()
    subscription.subscribe(ProjectionProbeSink(), projection, error)
    subscription.unsubscribe()
}

fun probeFallibleCall(error: CanonicalError): CanonicalError? =
    try {
        fallibleBindingProbe(shouldFail = true, error = error)
        null
    } catch (probeError: BindingProbeException.Canonical) {
        probeError.error
    }

suspend fun probeAsyncCall(ack: CommandAck): CommandAck = asyncBindingProbe(ack)

fun probeProjectionContracts(
    hostId: HostId,
    projectId: ProjectId,
    sessionId: SessionId,
    workflowId: WorkflowId,
    runtimeId: ProviderRuntimeId,
    cursor: Cursor,
    error: CanonicalError,
): List<ProjectionSubscribeAck> {
    val keys =
        listOf(
            ProjectionKey.ProjectIndex(hostId),
            ProjectionKey.SessionIndex(projectId),
            ProjectionKey.Transcript(sessionId),
            ProjectionKey.LiveActivity(sessionId),
            ProjectionKey.InputQueue(sessionId),
            ProjectionKey.AttentionInbox(hostId),
            ProjectionKey.WorkflowBoard(workflowId),
            ProjectionKey.RuntimeCapability(hostId, runtimeId),
        )
    val outcomes =
        listOf(
            ProjectionSubscribeOutcome.Resumed(cursor),
            ProjectionSubscribeOutcome.CurrentFollows(cursor),
            ProjectionSubscribeOutcome.Rejected(error),
        )

    return keys.mapIndexed { index, key ->
        val request = ProjectionSubscribe(key = key, since = cursor)
        consumeProjectionKey(request.key)
        ProjectionSubscribeAck(key = request.key, outcome = outcomes[index % outcomes.size]).also {
            consumeProjectionOutcome(it.outcome)
        }
    }
}

fun probeMobileIngress(
    deviceId: DeviceId,
    body: Command,
    contentRef: ContentRef,
    error: CanonicalError,
): String {
    val actor: Actor = Actor.Human(deviceId)
    val command = DeviceCommandRequest(idempotencyKey = "probe", ttlMs = 1uL, body = body)
    val write =
        ContentWriteRequest(
            contentKind = ContentKind.PLAIN_TEXT,
            byteLen = contentRef.byteLen,
            digest = contentRef.digest,
        )
    val responses: List<ContentWriteResponse> =
        listOf(ContentWriteResponse.Stored(contentRef), ContentWriteResponse.Rejected(error))

    return buildString {
        append(consumeActor(actor))
        append(command.idempotencyKey)
        append(write.contentKind.name)
        responses.forEach { append(consumeContentWriteResponse(it)) }
    }
}

fun probeRuntimeCapabilityScope(view: RuntimeCapabilityView): String =
    "${view.hostId.value}:${view.runtimeId.value}"

fun probeQuestionSet(prompt: QuestionPrompt, answer: QuestionAnswer, draft: MobileQuestionAnswer): String =
    "${prompt.questionKey}:${answer.questionKey}:${draft.questionKey}:${answer.optionIds.joinToString(",")}"

private fun consumeActor(actor: Actor): String =
    when (actor) {
        is Actor.Human -> actor.deviceId.value
        is Actor.Workflow -> actor.workflowId.value
        Actor.Broker -> "broker"
    }

private fun consumeProjectionKey(key: ProjectionKey): String =
    when (key) {
        is ProjectionKey.ProjectIndex -> key.hostId.value
        is ProjectionKey.SessionIndex -> key.projectId.value
        is ProjectionKey.Transcript -> key.sessionId.value
        is ProjectionKey.LiveActivity -> key.sessionId.value
        is ProjectionKey.InputQueue -> key.sessionId.value
        is ProjectionKey.AttentionInbox -> key.hostId.value
        is ProjectionKey.WorkflowBoard -> key.workflowId.value
        is ProjectionKey.RuntimeCapability -> "${key.hostId.value}:${key.runtimeId.value}"
    }

private fun consumeProjectionOutcome(outcome: ProjectionSubscribeOutcome): String =
    when (outcome) {
        is ProjectionSubscribeOutcome.Resumed -> outcome.fromCursor.seq.toString()
        is ProjectionSubscribeOutcome.CurrentFollows -> outcome.currentCursor.seq.toString()
        is ProjectionSubscribeOutcome.Rejected -> outcome.error.code.toString()
    }

private fun consumeContentWriteResponse(response: ContentWriteResponse): String =
    when (response) {
        is ContentWriteResponse.Stored -> response.contentRef.digest
        is ContentWriteResponse.Rejected -> response.error.code.toString()
    }
